apps/meteor/client/sidebar/header/MatrixFederationSearch/FederatedRoomList.tsx

```diff
@@ -0,0 +1,58 @@
+import { Throbber, Box } from '@rocket.chat/fuselage';
+import { useSetModal, useEndpoint } from '@rocket.chat/ui-contexts';
+import { useInfiniteQuery, useMutation } from '@tanstack/react-query';
+import type { VFC } from 'react';
+import React from 'react';
+import { Virtuoso } from 'react-virtuoso';
+
+import ScrollableContentWrapper from '../../../components/ScrollableContentWrapper';
+import FederatedRoomListItem from './FederatedRoomListItem';
+
+type FederatedRoomListProps = {
+	serverName: string;
+	roomName?: string;
+	pageToken?: string;
+	count?: number;
+};
+
+const FederatedRoomList: VFC<FederatedRoomListProps> = ({ serverName, roomName, count }) => {
+	const fetchRoomList = useEndpoint('GET', '/v1/federation/searchPublicRooms');
+	const joinExternalPublicRoom = useEndpoint('POST', '/v1/federation/joinExternalPublicRoom');
+
+	const setModal = useSetModal();
+	const { data, isLoading, isFetchingNextPage, fetchNextPage } = useInfiniteQuery(
+		['federation/searchPublicRooms', serverName, roomName, count],
+		async ({ pageParam }) => fetchRoomList({ serverName, roomName, count, pageToken: pageParam }),
+		{ getNextPageParam: (lastPage) => lastPage.nextPageToken },
+	);
+
+	const { mutate: onClickJoin, isLoading: isLoadingMutation } = useMutation(
+		['federation/joinExternalPublicRoom'],
+		async (id: string) => {
+			return joinExternalPublicRoom({ externalRoomId: id as `!${string}:${string}` });
+		},
+		{ onSuccess: () => setModal(null) },
+	);
```

Comment:
Maybe create two different hooks, one for the query and other for the Mutation, what do you think? 
That way these hooks can be used elsewhere